Error on linking: undefined reference to `vtable for Serials::AbstractSerial'

ino.cpp.o:(.text._ZN7Serials16InternalSoftwareD2Ev[Serials::InternalSoftware::~InternalSoftware()]+0x0): undefined reference to `vtable for Serials::AbstractSerial'
collect2: error: ld returned 1 exit status

--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,5 @@
 src/main.cpp
+src/src.ino
 platformio.ini
 .pioenvs
 .piolibdeps

--- a/library.json
+++ b/library.json
@@ -1,6 +1,6 @@
 {
   "name": "Nova Fitness Sds dust sensors library",
-  "version": "1.3.0",
+  "version": "1.3.2",
   "keywords": "sds011, sds, sds021, dust, sensor, pm10, pm25, arduino, esp8266",
   "description": "A high-level abstaction over Sds sensors family",
   "homepage": "https://github.com/lewapek/sds-dust-sensors-arduino-library",

--- a/library.properties
+++ b/library.properties
@@ -1,5 +1,5 @@
 name=Nova Fitness Sds dust sensors library
-version=1.3.0
+version=1.3.2
 author=Paweł Kołodziejczyk <lewapek@gmail.com>
 maintainer=Pawel Kołodziejczyk <lewapek@gmail.com>
 sentence=A high-level abstaction over Sds sensors family

--- a/src/Serials.h
+++ b/src/Serials.h
@@ -6,10 +6,18 @@
 
 namespace Serials {
 
-  struct AbstractSerial {
-    virtual void begin(int baudRate) ;
-    virtual Stream *getStream();
-    virtual ~AbstractSerial() {}
+  // cpp is really weird...
+  // there was compilation warning about missing virtual destructor for class AbstractSerial
+  // I added virtual destructor with empty body
+  // remaining virtual methods didn't have body (they were 'pure virtual')
+  // after that the following error appeared: undefined reference to `vtable for Serials::AbstractSerial'
+  // what a nonsense ...
+  // just to satisfy linker in gcc I needed to add empty parentheses to other virtual methods...
+  class AbstractSerial {
+  public:
+    virtual void begin(int baudRate) {};
+    virtual Stream *getStream() {};
+    virtual ~AbstractSerial() {};
   };
 
   struct Hardware: public AbstractSerial {
@@ -42,17 +50,23 @@ namespace Serials {
 
   struct InternalSoftware: public AbstractSerial {
     InternalSoftware(const int &pinRx, const int &pinTx):
-      serial(SoftwareSerial(pinRx, pinTx)) {}
+      serial(new SoftwareSerial(pinRx, pinTx)) {}
+
+    ~InternalSoftware() {
+      if (serial != NULL) {
+        delete serial;
+      }
+    }
 
     void begin(int baudRate) {
-      serial.begin(baudRate);
+      serial->begin(baudRate);
     }
 
     Stream *getStream() {
-      return &serial;
+      return serial;
     }
 
-    SoftwareSerial serial;
+    SoftwareSerial *serial;
   };
 
 }

